nntrainer/engine.cpp

```diff
@@ -0,0 +1,181 @@
+
+std::mutex engine_mutex;
+
+std::once_flag global_engine_init_flag;
+
+void Engine::add_default_object(Engine &eg) {
+  /// @note all layers should be added to the app_context to guarantee that
+  /// createLayer/createOptimizer class is created
+
+  nntrainer::AppContext *app_context = new nntrainer::AppContext();
+  app_context->Global();
+
+  eg.registerContext("cpu", app_context);
+
+  // #ifdef ENALBE_OPENCL
+  //   eg.registererContext("gpu",
+  //                        nntrainer::ClContext(nntrainer::ClContext::Global()));
+  // #endif
+}
+
+void Engine::registerer(Engine &eg) noexcept {
+  try {
+    add_default_object(eg);
+  } catch (std::exception &e) {
+    ml_loge("registering layers failed!!, reason: %s", e.what());
+  } catch (...) {
+    ml_loge("registering layer failed due to unknown reason");
+  }
+};
+
+Engine &Engine::Global() {
+  static Engine instance;
+  /// in g++ there is a bug that hangs up if caller throws,
+  /// so registerer is noexcept although it'd better not
+  /// https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70298
+  std::call_once(global_engine_init_flag, registerer, std::ref(instance));
+  return instance;
+}
+
+std::string
+Engine::parseComputeEngine(const std::vector<std::string> &props) const {
+  for (auto &prop : props) {
+    std::string key, value;
+    int status = nntrainer::getKeyValue(prop, key, value);
+    if (nntrainer::istrequal(key, "engine")) {
+      constexpr const auto data =
+        std::data(props::ComputeEngineTypeInfo::EnumList);
+      for (uint i = 0; i < props::ComputeEngineTypeInfo::EnumList.size(); ++i) {
+        if (nntrainer::istrequal(value.c_str(),
+                                 props::ComputeEngineTypeInfo::EnumStr[i])) {
+          return props::ComputeEngineTypeInfo::EnumStr[i];
+        }
+      }
+    }
+  }
+
+  return "cpu";
+}
+
+/**
+ * @brief Get the Full Path from given string
+ * @details path is resolved in the following order
+ * 1) if @a path is absolute, return path
+ * ----------------------------------------
+ * 2) if @a base == "" && @a path == "", return "."
+ * 3) if @a base == "" && @a path != "", return @a path
+ * 4) if @a base != "" && @a path == "", return @a base
+ * 5) if @a base != "" && @a path != "", return @a base + "/" + path
+ *
+ * @param path path to calculate from base
+ * @param base base path
+ * @return const std::string
+ */
+const std::string getFullPath(const std::string &path,
+                              const std::string &base) {
+  /// if path is absolute, return path
+  if (path[0] == '/') {
+    return path;
+  }
+
+  if (base == std::string()) {
+    return path == std::string() ? "." : path;
+  }
+
+  return path == std::string() ? base : base + "/" + path;
+}
+
+const std::string Engine::getWorkingPath(const std::string &path) {
+  return getFullPath(path, working_path_base);
+}
+
+void Engine::setWorkingDirectory(const std::string &base) {
+  DIR *dir = opendir(base.c_str());
+
+  if (!dir) {
```

Comment:
`if (!dir && errno == EPERM)`

Other issue might be that file descriptor limit per user/process/cgroup was reached.
Then `errno` would be `EMFILE` or `ENFILE`.
User getting such error for that case would be really puzzled/misguided, I imagine and would take probably `strace/gdb` to solve what is really going on.